### PR TITLE
MQTT: fix event flow & add more tests

### PR DESCRIPTION
**Pull request by lolBig**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added

### CODE REVIEW

**Comment by lolBig** [comment]
issue: #548

File: docs/api/MQTT.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ client.on('connect', function () {
 client.on('message', function (topic, message) {
   // message is Buffer
   console.log(message.toString());
-  client.end();
+  client.disconnect();
```
Comment (yorkie) [inline docs/api/MQTT.md]: The MQTT API keeps consistent with [MQTT.js](https://github.com/mqttjs/MQTT.js), which doesn't use `disconnect()`.

File: docs/api/MQTT.md [hunk 2]
```diff
@@ -111,10 +111,13 @@ Unsubscribe from a topic.
 * `force` {Boolean} passing it to true will close the client right away.
 * `callback` {Function} will be called when the client is closed. This parameter is optional.
 
-Close the client.
+### client.disconnect([err, [, callback]])
```
Comment (yorkie) [inline docs/api/MQTT.md]: Ditto

File: docs/api/MQTT.md [hunk 3]
```diff
@@ -111,10 +111,13 @@ Unsubscribe from a topic.
...
-### client.reconnect()
+* `err` {Error} throw an error before disconnect. This parameter is optional.
+* `callback` {Function} will be called when the client is closed. This parameter is optional.
 
-Connect again using the same options as connect().
+Disconnect the client right away The `client` can safety be unref after call this method.
+
+### client.reconnect()
 
-For more detailed, please see [MQTT.js Readme](https://github.com/mqttjs/MQTT.js#client).
```
Comment (yorkie) [inline docs/api/MQTT.md]: As I have commented, this rule should not be removed from this documentation.

**Comment by yorkie** [comment]
BTW, if a PR branch is to fix a specific issue, a suggestion name shall be like `fix/gh-648`.

File: docs/api/MQTT.md [hunk 4]
```diff
@@ -113,8 +113,17 @@ Unsubscribe from a topic.
 
 Close the client.
 
+### client.disconnect([err, [, callback]])
```
Comment (legendecas) [inline docs/api/MQTT.md]: What's the difference between `client.end` and `client.disconnect`?

Comment (lolBig) [reply docs/api/MQTT.md]: they both do the same thing, the MQTT.js use `end` and we use `disconnect` before, so keep `disconnect` and add `end` to make compatible with `MQTT.js` and old apps

**Pull request merged**